Agent/Public/NewRelic.m

```diff
@@ -600,10 +601,53 @@ + (BOOL) incrementAttribute:(NSString*)name
                                                                                       persistent:YES];
 }
 
-+ (BOOL) setUserId:(NSString*)userId {
-    return [[NewRelicAgentInternal sharedInstance].analyticsController setSessionAttribute:@"userId"
-                                                                                     value:userId
-                                                                                persistent:YES];
++ (BOOL) setUserId:(NSString* _Nullable)userId {
+    NSString *previousUserId = [[NewRelicAgentInternal sharedInstance] getUserId];
+    BOOL newSession = false;
+    // If the client passes a new userId that is non NULL.
+    if (userId != NULL) {
+
+        // A new userId has been set where the previously set one (during this app session (since app launch))was not NULL.
+
+        if (previousUserId != NULL) {
+            // continue session and set the new userID
+            newSession = true;
+        }
+        // If userId has not previously been set this session
+        else {
+            newSession = true;
```

Comment:
I think this else is not needed.

Reply:
Thanks, I pushed a commit that rectifies this issue by always starting a new Session if the passed userId is non Null.